language/move-lang/src/hlir/translate.rs

```diff
@@ -1415,6 +1421,101 @@ fn bind_for_short_circuit_sequence(seq: &T::Sequence) -> bool {
         }
 }
 
+//**************************************************************************************************
+// Trailing semicolon
+//**************************************************************************************************
+
+fn check_trailing_unit(context: &mut Context, block: &mut Block) {
+    use H::{Command_ as C, Statement_ as S, UnannotatedExp_ as E};
+    macro_rules! hcmd {
+        ($loc:pat, $cmd:pat) => {
+            sp!(_, S::Command(sp!($loc, $cmd)))
+        };
+    }
+    macro_rules! hexp {
+        ($loc:pat, $e:pat) => {
+            hcmd!(_, C::IgnoreAndPop { exp: H::Exp { exp: sp!($loc, $e), .. }, .. })
+        };
+    }
+    fn divergent_block(block: &Block) -> bool {
+        matches!(
+            block.back(),
+            Some(hcmd!(_, C::Break))
+                | Some(hcmd!(_, C::Continue))
+                | Some(hcmd!(_, C::Abort(_)))
+                | Some(hcmd!(_, C::Return(_)))
+                | Some(hexp!(_, E::Unreachable))
+        )
+    }
+    macro_rules! invalid_trailing_unit {
+        ($context:ident, $loc:expr, $uloc:expr) => {{
+            let semi_msg = "Unnecessary trailing ';'";
+            let unreachable_msg = "Divergent expression. Any code following the evaluation of \
```

Comment:
I think we should try to improve this message.

- "Unnecessary" seems like the wrong word. This isn't just a suggestion to simplify the code; we're treating this as invalid. So maybe change it to "Invalid"?
- The unreachable_msg has too much jargon, specifically "divergent" and "dead". Maybe drop "Divergent expression" and replace the whole thing with: "Any code after this expression will not be reached"?
- The info_msg seems pretty good as-is